src/deploy/functions/release/sourceTokenScraper.ts

```diff
@@ -1,37 +1,64 @@
+import { FirebaseError } from "../../../error";
 import { logger } from "../../../logger";
 
+type TokenFetchState = "NONE" | "FETCHING" | "VALID";
+
 /**
  * GCF v1 deploys support reusing a build between function deploys.
  * This class will return a resolved promise for its first call to tokenPromise()
  * and then will always return a promise that is resolved by the poller function.
  */
 export class SourceTokenScraper {
-  private firstCall = true;
-  private resolve!: (token: string) => void;
+  private tokenValidDurationMs;
+  private resolve!: (token?: string) => void;
   private promise: Promise<string | undefined>;
+  private expiry: number | undefined;
+  private fetchState: TokenFetchState;
 
-  constructor() {
+  constructor(validDurationMs = 1500000) {
+    this.tokenValidDurationMs = validDurationMs;
     this.promise = new Promise((resolve) => (this.resolve = resolve));
+    this.fetchState = "NONE";
+  }
+
+  async getToken(): Promise<string | undefined> {
+    if (this.fetchState === "NONE") {
+      this.fetchState = "FETCHING";
+      return undefined;
+    } else if (this.fetchState === "FETCHING") {
+      return this.promise; // wait until we get a source token
+    } else if (this.fetchState === "VALID") {
+      if (this.isTokenExpired()) {
+        this.fetchState = "FETCHING";
+        this.promise = new Promise((resolve) => (this.resolve = resolve));
+        return undefined;
+      }
+      return this.promise;
+    }
   }
 
-  // Token Promise will return undefined for the first caller
-  // (because we presume it's this function's source token we'll scrape)
-  // and then returns the promise generated from the first function's onCall
-  tokenPromise(): Promise<string | undefined> {
-    if (this.firstCall) {
-      this.firstCall = false;
-      return Promise.resolve(undefined);
+  isTokenExpired(): boolean {
+    if (this.expiry === undefined) {
+      throw new FirebaseError("failed to check expiry: no token exists");
```

Comment:
This error message makes sense to you and me but won't to almost all users trying to deploy their function.

I think this is a good example of a user facing error message - describe the situation in coarsely and suggest next steps:

https://github.com/firebase/firebase-tools/blob/2ebefe2d8c5e33e9fc8c078235d02579ccdbf036/src/deploy/functions/prepareFunctionsUpload.ts#L36-L40

In our case, hitting this if statement should never happen and would be considered a bug. Maybe we can suggest that they file a issue.

Reply:
ah gotcha. working on a more descriptive message.

Reply:
```
throw new FirebaseError(
        "Your deployment is checking the expiration of a source token that has not yet been polled. " +
          "Hitting this case should never happen and should be considered a bug. " +
          "Please file an issue at https://github.com/firebase/firebase-tools/issues"
      );
```
What about something like this?